jena-fuseki2/jena-fuseki-core/src/main/java/org/apache/jena/fuseki/build/RefCountingMap.java

```diff
@@ -0,0 +1,187 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.fuseki.build;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * A key -> value 'map' which reference counts entries.  
+ * 
+ * <p>
+ *   The same (key,value) pair can be added to the map several times and then 
+ *   removed several times.  A reference count is incremented for each addition
+ *   and, provided the count is greater than 0, decremented on removal.
+ * <p>
+ * 
+ * <p>
+ *   The pair is removed from the map when a remove decrements the reference count to 0.
+ * </p>
+ * 
+ * <p>
+ *   This class is thread safe.
+ * </p>
+ * 
+ * @param <K>
+ * @param <T>
+ */
+public class RefCountingMap<K, T> {
```

Comment:
There's, understandably, a lot of methods here that just pass-through to `ConcurrentHashMap`. Maybe it would be worth doing `extends ConcurrentHashMap<K, CountedRef<T>>` and just overriding the methods that need overriding?